fourier/fourier_sincos.go

```diff
@@ -0,0 +1,238 @@
+//
+// If the length of src is not t.Len(), Transform will panic.
+// If dst is nil, a new slice is allocated and returned. If dst is not nil and
+// the length of dst does not equal t.Len(), FFT will panic.
+// It is safe to use the same slice for dst and src.
+func (t *DCT) Transform(dst, src []float64) []float64 {
+	if len(src) != t.Len() {
+		panic("fourier: sequence length mismatch")
+	}
+	if dst == nil {
+		dst = make([]float64, t.Len())
+	} else if len(dst) != t.Len() {
+		panic("fourier: destination length mismatch")
+	}
+	copy(dst, src)
+	fftpack.Cost(len(dst), dst, t.work, t.ifac[:])
+	return dst
+}
+
+// DST implements Discrete Sine Transform for real sequences.
+type DST struct {
+	work []float64
+	ifac [15]int
+}
+
+// NewDST returns a DST initialized for work on sequences of length n.
+func NewDST(n int) *DST {
+	var t DST
+	t.Reset(n)
+	return &t
+}
+
+// Len returns the length of the acceptable input.
+func (t *DST) Len() int { return (2*len(t.work)+1)/5 - 1 }
+
+// Reset reinitializes the DCT for work on sequences of length n.
+func (t *DST) Reset(n int) {
+	if 5*(n+1)/2 <= cap(t.work) {
+		t.work = t.work[:5*(n+1)/2]
+	} else {
+		t.work = make([]float64, 5*(n+1)/2)
+	}
+	fftpack.Sinti(n, t.work, t.ifac[:])
+}
+
+// Transform computes the Discrete Fourier Sine Transform of the input
+// data, src, placing the result in dst and returning it.
+// This transform is unnormalized; a call to Transform followed by
+// another call to Transform will multiply the input sequence by 2*(n-1),
+// where n is the length of the sequence.
+//
+// If the length of src is not t.Len(), Transform will panic.
+// If dst is nil, a new slice is allocated and returned. If dst is not nil and
+// the length of dst does not equal t.Len(), FFT will panic.
+// It is safe to use the same slice for dst and src.
+func (t *DST) Transform(dst, src []float64) []float64 {
+	if len(src) != t.Len() {
+		panic("fourier: sequence length mismatch")
+	}
+	if dst == nil {
+		dst = make([]float64, t.Len())
+	} else if len(dst) != t.Len() {
+		panic("fourier: destination length mismatch")
+	}
+	copy(dst, src)
+	fftpack.Sint(len(dst), dst, t.work, t.ifac[:])
+	return dst
+}
+
+// QuarterWaveFFT implements Fast Fourier Transform for quarter wave data.
+type QuarterWaveFFT struct {
+	work []float64
+	ifac [15]int
+}
+
+// NewQuarterWaveFFT returns a QuarterWaveFFT initialized for work on sequences of length n.
+func NewQuarterWaveFFT(n int) *QuarterWaveFFT {
+	var t QuarterWaveFFT
+	t.Reset(n)
+	return &t
+}
+
+// Len returns the length of the acceptable input.
+func (t *QuarterWaveFFT) Len() int { return len(t.work) / 3 }
+
+// Reset reinitializes the QuarterWaveFFT for work on sequences of length n.
+func (t *QuarterWaveFFT) Reset(n int) {
+	if 3*n <= cap(t.work) {
+		t.work = t.work[:3*n]
+	} else {
+		t.work = make([]float64, 3*n)
+	}
+	fftpack.Cosqi(n, t.work, t.ifac[:])
+}
+
+// CosCoefficients computes the Fast Fourier Transform of quarter wave data for
```

Comment:
Should this say Discrete? DST for instance uses discrete while this uses fast.

Reply:
They are sort of the same in this domain. The docs reflect the documentation in fftpack.